### ISSUE TITLE
etcdserver/security: What is the expected behavior when waiting for changes? Can currently see changes that shouldn't be visible.

### ISSUE THREAD

**Issue by tomdee** [coreos/etcd#2555]
I've just been trying out the new security API and found some unexpected behavior when using `?wait=true&recursive=true`

What is the expected behavior here? It's not documented currently in the RFC. e.g. will curl return even if there are no visible changes? If so, what does the JSON look like.

The unexpected behavior that I observed on the current latest code

1. I enabled security
2. I created a user (user1) and assigned them a role (role1)
3. I created a directory (/dir)
4. I gave the role permissions to read and write /dir and /dir/allowed
5. I verified that the user could write a key /dir/allowed but not write a key /dir/notallowed
6. I did a read waiting for changes recursively on /dir as the user
```
curl --user user1:pass 'http://127.0.0.1:2379/v2/keys/dir?wait=true&recursive=true'
{"action":"set","node":{"key":"/dir/root","value":"Hello world","modifiedIndex":14,"createdIndex":14}}

curl --user user1:pass http://127.0.0.1:2379/v2/keys/dir/root
{"message":"Insufficient credentials"}
```
When I wrote to /dir/root using the root user, I observed that my user was able to see this change (even though they can't read that key directly)

**Comment by tomdee** [comment]
@kelseyhightower this is the issue that @apollitt mentioned when you were chatting the other day.

**Comment by xiang90** [comment]
@barakmich This is fixed, right? I cannot reproduce this.

**Comment by tomdee** [comment]
I don't time to retest at the moment, but I've been told that this is fixed so I'm happy to just close this issue.

**Comment by barakmich** [comment]
Yeah; I believe this was one of the many fixes in #2764
